### PR TITLE
Improve tolerance for ending scroll animations

### LINKED ISSUE #828: Janky tween in the shopping list demo
In the [shopping list demo](https://gist.github.com/appsforartists/ad38541cd05098018f55), if you scroll the list (which is smaller than its container) and release, there appears to be some jank towards the end of the animation back to its initial position.

It looks like the tween gets close, and then skips the last frames.  It feels like this:

…"slow but smooth", "slow but smooth", "slow but smooth", "slow but smooth", "snap!"

and the snap is jarring.

Happy to show you in-person if this doesn't make any sense.

### PR DESCRIPTION
**Pull request by abarth**
We had the units wrong on the tolerances. Previously we multiplied by the
device pixel ratio, which meant we got larger tolerances as we got more
resolution. Also, simplify logic in Newton for applying the tolerances.

Fixes #828

### CODE REVIEW

**Comment by abarth** [comment]
@chinmaygarde @HansMuller

File: packages/flutter/lib/src/animation/scroll_behavior.dart [hunk 1]
```diff
@@ -105,26 +105,25 @@ class UnboundedBehavior extends ExtentScrollBehavior {
...
 
-  // Assume that we're rendering at atleast 15 FPS. Stop when we're
-  // scrolling less than one logical pixel per frame. We're essentially
-  // normalizing by the devicePixelRatio so that the threshold has the
-  // same effect independent of the device's pixel density.
-  double endVelocity = 15.0 * ui.window.devicePixelRatio;
+  // Assume that we're rendering at >= 15 FPS. Stop when we're
+  // scrolling less than one physical pixel per frame, which isn't
+  // visually detectable.
+  double endVelocity = 15.0 / ui.window.devicePixelRatio;
```
Comment (Hixie) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: I would rephrase this. It's not that we're assuming that we're getting at least 15fps. It's "Assume that if we move less than one device pixel in 66ms, the user won't notice if we just snap the animation to the end and call it a day".

...and I think 66ms is probably too short. I think we should stop only when we reach about 1 device pixel per 500ms.

I think I would set this to `1 / (0.500 * ui.window.devicePixelRatio) // logical pixels per second`

Anyway. I don't feel strongly about the value, but either way please do per the units in a comment.

File: packages/flutter/lib/src/animation/scroll_behavior.dart [hunk 2]
```diff
@@ -105,26 +105,25 @@ class UnboundedBehavior extends ExtentScrollBehavior {
...
 
   // Similar to endVelocity. Stop scrolling when we're this close to
   // destiniation scroll offset.
-  double endDistance = 0.5 * ui.window.devicePixelRatio;
+  double endDistance = 1.0 / ui.window.devicePixelRatio;
```
Comment (Hixie) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: needs units in comments (`// logical pixels`)

File: packages/flutter/lib/src/animation/scroll_behavior.dart [hunk 3]
```diff
@@ -105,26 +105,25 @@ class UnboundedBehavior extends ExtentScrollBehavior {
 Simulation _createFlingScrollSimulation(double position, double velocity, double minScrollOffset, double maxScrollOffset) {
   double startVelocity = velocity * _kSecondsPerMillisecond;
```
Comment (Hixie) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: All these "end" values should be prefixed with a comment about how if you have a transform scaling this stuff up, our thresholds are going to make no sense and it's going to get ugly.

File: packages/flutter/lib/src/animation/scroll_behavior.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: The problem with choosing a large value like 500ms is that we don't fire the `onScrollEnd` callback until we finish the animations, which means there's a cost to coasting down to a tiny velocity.

File: packages/flutter/lib/src/animation/simulation_stepper.dart [hunk 4]
```diff
@@ -97,7 +97,10 @@ class SimulationStepper {
 
   void _tick(Duration elapsed) {
     double elapsedInSeconds = elapsed.inMicroseconds.toDouble() / Duration.MICROSECONDS_PER_SECOND;
-    _value = _simulation.x(elapsedInSeconds);
+    double x = _simulation.x(elapsedInSeconds);
+    double dx = _simulation.dx(elapsedInSeconds);
+    print('x=$x dx=$dx delta=${x - _value} _simulation=${_simulation.runtimeType}');
+    _value = x;
```
Comment (abarth) [inline packages/flutter/lib/src/animation/simulation_stepper.dart]: This change is spurious.

File: packages/flutter/lib/src/animation/scroll_behavior.dart [hunk 5]
```diff
@@ -105,26 +105,25 @@ class UnboundedBehavior extends ExtentScrollBehavior {
...
       new ScrollSimulation(position, startVelocity, minScrollOffset, maxScrollOffset, spring, _kScrollDrag)
-    ..tolerance = new Tolerance(velocity: endVelocity.abs(), distance: endDistance);
+    ..tolerance = new Tolerance(velocity: endVelocity, distance: endDistance);
   return simulation;
 }
 
 Simulation _createSnapScrollSimulation(double startOffset, double endOffset, double velocity) {
   double startVelocity = velocity * _kSecondsPerMillisecond;
-  double endVelocity = 15.0 * ui.window.devicePixelRatio * velocity.sign;
+  double endVelocity = velocity.sign * 15.0 / ui.window.devicePixelRatio;
```
Comment (Hixie) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: This shouldn't be hard-coded.

I think we should move all the tolerances into a class at the top level of this file, something like:

```dart
class Tolerances {
  static double endVelocityTolerance = ...; // logical pixels per second
  static double endDistanceTolerance = ...; // logical pixels
}
```

...that way if people are fiddling with transforms they can easily twiddle the values for all the classes and functions here instead of having to do it one by one (or not being able to easily change it at all).

File: packages/flutter/lib/src/animation/simulation_stepper.dart [hunk 6]
```diff
@@ -97,7 +97,10 @@ class SimulationStepper {
 
   void _tick(Duration elapsed) {
     double elapsedInSeconds = elapsed.inMicroseconds.toDouble() / Duration.MICROSECONDS_PER_SECOND;
-    _value = _simulation.x(elapsedInSeconds);
+    double x = _simulation.x(elapsedInSeconds);
+    double dx = _simulation.dx(elapsedInSeconds);
+    print('x=$x dx=$dx delta=${x - _value} _simulation=${_simulation.runtimeType}');
```
Comment (Hixie) [inline packages/flutter/lib/src/animation/simulation_stepper.dart]: print

Comment (Hixie) [inline packages/flutter/lib/src/animation/simulation_stepper.dart]: in fact this whole change looks like it was just a debugging-time thing.

File: packages/flutter/lib/src/animation/scroll_behavior.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: what do we do onScrollEnd?

**Comment by Hixie** [comment]
LGTM.

File: packages/flutter/lib/src/animation/simulation_stepper.dart (hunk 6 above)
Comment (abarth) [inline packages/flutter/lib/src/animation/simulation_stepper.dart]: Yeah, I thought I removed it before upload but failed.  :)

File: packages/flutter/lib/src/animation/scroll_behavior.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: We start fading out the scrollbars, for example.

Comment (Hixie) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: Fair enough.

In any case I think it should be written in terms of time, not frequency. We care about the time it takes for the scroll bar to fade out, not the number of frames. If we want it to be 66ms, then it'd be `1 / (0.066 * ui.window.devicePixelRatio) // logical pixels per second)`. Or if you want the dimension analysis to be more obvious, `(1 / ui.window.devicePixelRatio) / (0.066)` maybe.

File: packages/flutter/lib/src/animation/scroll_behavior.dart (hunk 5 above)
Comment (abarth) [inline packages/flutter/lib/src/animation/scroll_behavior.dart]: Done (slightly differently).
